### ISSUE TITLE
Postgresql Integrations

### ISSUE THREAD

**Issue by nsunmaz** [borisdj/EFCore.BulkExtensions#67]
I use "**Npgsql.EntityFrameworkCore.PostgreSQL**" with EF Core in the project. Does your library support **PostgreSQL**? If not, do you have a plan to do this later?

**Comment by neco122** [comment]
Thanks for information.

**Comment by collinbarrett** [comment]
Looks like a great library! I am also not using SQL Server, though. I'd recommend making the db provider support more prominent on the README, as it is not immediately obvious to potential users that it only supports SQL Server.

**Comment by borisdj** [comment]
It is stated at the end of first paragraph:

> Under the hood uses SqlBulkCopy for Insert, for Update/Delete combines BulkInsert with raw Sql MERGE (MsSQL 2008+).

**Comment by richard-s-popov** [comment]
Still no news about support PostgreSQL?

**Comment by borisdj** [comment]
Not yet.
Have managed to extend it for SQLite recently.
PostgreSql maybe sometimes during next year.

**Comment by MaceWindu** [comment]
As workaround you can use linq2db [BulkCopy](https://github.com/linq2db/linq2db#bulk-copy) functionality through [EF.Core adapter](https://github.com/linq2db/linq2db.EntityFrameworkCore)

https://github.com/linq2db/linq2db/wiki/Bulk-Copy#postgresql-provider-specific-bulk-copy

**Comment by IvanAntipov** [comment]
EFCore.BuildExtension is a great library, unfortunately it doesn't support Postgres. You can use 'FlexLabs.Upsert' nuget as alternative for Postgres

**Comment by etiennecl** [comment]
Is this still in the roadmap? Anything I could do to help? I'm not an expert but I sure could give you a hand.

**Comment by borisdj** [comment]
Yes but still no timeframe.
There is one PR a few days ago in this direction: https://github.com/borisdj/EFCore.BulkExtensions/pull/434

Here are some links (Docs) that I've found when did the research:

https://www.npgsql.org/doc/copy.html
https://www.postgresql.org/docs/9.2/static/sql-copy.html
https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues/113
__
https://www.bytefish.de/blog/postgresql_bulk_insert/
https://dba.stackexchange.com/questions/41059/optimizing-bulk-update-performance-in-postgresql
http://thebuild.com/blog/2016/01/14/doing-a-bulk-merge-in-postgresql-9-5/
https://www.alibabacloud.com/blog/how-does-postgresql-implement-batch-update-deletion-and-insertion_596030

https://stackoverflow.com/questions/30320418/insert-the-whole-value-of-datatable-bulk-into-postgresql-table
https://stackoverflow.com/questions/758945/whats-the-fastest-way-to-do-a-bulk-insert-into-postgres
https://stackoverflow.com/questions/5940225/fastest-way-of-inserting-in-entity-framework
https://stackoverflow.com/questions/43531602/bulk-update-in-entity-framework-core/43531693

**Comment by borisgr04** [comment]
> Sí, pero aún sin plazo.
> Hace unos días hay un RP en esta dirección: [# 434](https://github.com/borisdj/EFCore.BulkExtensions/pull/434)
> 
> Aquí hay algunos enlaces (Documentos) que encontré cuando hice la investigación:
> 
> [https://www.npgsql.org/doc/copy.html ](https://www.npgsql.org/doc/copy.html)
> [https://www.postgresql.org/docs/9.2/static/sql-copy.html ](https://www.postgresql.org/docs/9.2/static/sql-copy.html)
> [npgsql / efcore.pg # 113](https://github.com/npgsql/efcore.pg/issues/113)
> __
> [https: // www. bytefish.de/blog/postgresql_bulk_insert/ ](https://www.bytefish.de/blog/postgresql_bulk_insert/)
> [https://dba.stackexchange.com/questions/41059/optimizing-bulk-update-performance-in-postgresql ](https://dba.stackexchange.com/questions/41059/optimizing-bulk-update-performance-in-postgresql)
> [http://thebuild.com/blog/2016/01/14/doing -una-fusión-masiva-en-postgresql-9-5 / ](http://thebuild.com/blog/2016/01/14/doing-a-bulk-merge-in-postgresql-9-5/)
> https://www.alibabacloud.com/blog/how-does-postgresql-implement-batch-update-deletion-and-insertion_596030
> 
> [https://stackoverflow.com/questions/30320418/insert-the-whole-value-of-datatable-bulk-into-postgresql-table ](https://stackoverflow.com/questions/30320418/insert-the-whole-value-of-datatable-bulk-into-postgresql-table)
> [https://stackoverflow.com/questions/758945/whats-the-fastest-way -to-do-a-bulk-insert-into-postgres ](https://stackoverflow.com/questions/758945/whats-the-fastest-way-to-do-a-bulk-insert-into-postgres)
> [https://stackoverflow.com/questions/5940225/fastest-way-of-inserting-in-entity-framework ](https://stackoverflow.com/questions/5940225/fastest-way-of-inserting-in-entity-framework)
> [https://stackoverflow.com/questions/43531602 / actualización-masiva-en-entidad-framework-core / 43531693](https://stackoverflow.com/questions/43531602/bulk-update-in-entity-framework-core/43531693)

I want help. How?

**Comment by borisdj** [comment]
I started some work on this, made so far shell of methods and prof of concept for basic Insert.
Once I commit it, then next phase would be implementing each method.

**Comment by ahmeetarif** [comment]
WE ALL ARE WAITING FOR POSTGRESQL SUPPORT :)

**Comment by aktxyz** [comment]
+1

**Comment by douglasg14b** [comment]
This would be very nice to have, just switched to Postgres and was quite surprised when the queries stopped working :/

**Comment by trojanmartin** [comment]
> I started some work on this, made so far shell of methods and prof of concept for basic Insert. Once I commit it, then next phase would be implementing each method.

Can u push that branch into repo? We will be happy to help with the implementation, if possible.

**Comment by borisdj** [comment]
PostgreSql adapter added with with v6.0.4
Currently supported methods: Insert and Upsert.
Still initial faze so complex cases might give exceptions.

**Comment by borisdj** [comment]
PostgreSql now fully supported, all Bulk and Batch ops, v6.0.5

**Comment by dimkadv2** [comment]
Hi!

sorry if I'm writing to the wrong thread

i faced the following problems while working PostgreSql:
when using json columns (in my case it's List<T> property) i have got error: 
```
The CLR type Ocs.API.Infrastructure.ResponseData.CatalogPathData isn't natively supported by Npgsql or your PostgreSQL.
To use it with a PostgreSQL composite you need to specify DataTypeName or to map it, please refer to the documentation
```

and i have problems with enums:

```c#

    public enum ConditionType : byte
    {
        Regular,
        Sale,
        Unconditional
    }


    public ConditionType Condition { get; set; }

```
error:
```
A PostgreSQL type with the name condition_type was not found in the database
```

P.S. i'm using EFCore.NamingConventions package and EFCore 6

thank you in advance!

**Comment by bugrakosen** [comment]
Hi,

I'm facing the same issue as @dimkadv2 with enum types.

It seems @Hoffs solution did not change anything for my situation.

Do you have any suggestions for me?

**Comment by borisdj** [comment]
Enums should be fixed with v6.1.8

**Comment by Bouke** [comment]
How invasive is it to backport this to support EFCore 3.1? We are running on .NET Framework which is only supported by EFCore 3.1.
